### PR TITLE
Implement digging feature

### PR DESCRIPTION
**Pull request by Sieluna**
~~Currently in the testing phase; please do not merge this branches.~~

Since I’m not very familiar with the interfaces and specific processes, I hope the coding process can be more transparent so I can get your thoughts on the interface design in a timely manner. This will help reduce potential refactoring costs later on.

### CODE REVIEW

**Comment by Sieluna** [comment]
By the way, could you let me know which part might be causing the loot drop to be so huge?
vlcsnap-2024-11-06-16h16m52s301

**Comment by DevBobcorn** [comment]
Just wanted to make them easier seen from topdown view :P

**Comment by DevBobcorn** [comment]
The entity mesh is created here https://github.com/DevBobcorn/CornCraft/blob/master/Assets/Scripts/Rendering/Entity/ItemEntityRender.cs#L28-L40

File: Assets/Scripts/Client/BaseCornClient.cs [hunk 1]
```diff
@@ -220,7 +222,7 @@ protected virtual void SetWorldOriginOffset(Vector3Int offset)
         public abstract bool UseItemOnOffHand();
         public abstract bool DoWindowAction(int windowId, int slotId, WindowActionType action);
         public abstract bool PlaceBlock(BlockLoc blockLoc, Direction blockFace, Hand hand = Hand.MainHand);
-        public abstract bool DigBlock(BlockLoc blockLoc, bool swingArms = true, bool lookAtBlock = true);
+        public abstract float DigBlock(BlockLoc blockLoc, Direction blockFace, DiggingStatus status = DiggingStatus.Started, bool swingArms = true, bool lookAtBlock = true);
```
Comment (Sieluna) [inline Assets/Scripts/Client/BaseCornClient.cs:225]: For this part of the code, do you have any good suggestions? I feel it makes the code messy.

Comment (Sieluna) [reply Assets/Scripts/Client/BaseCornClient.cs:225]: Actually, I could move the code elsewhere, but I don't want the scripts to reference each other.

Comment (DevBobcorn) [reply Assets/Scripts/Client/BaseCornClient.cs:225]: I think this part is fine, but maybe it's a good idea to put [tool digging effectiveness](https://github.com/DevBobcorn/CornCraft/pull/20/files#diff-e9f44e86b893676a15c40814f4bd69a3db533a94455ca51f114e4e7aebef2c0c) into item type definition or item type palette.

Comment (DevBobcorn) [reply Assets/Scripts/Client/BaseCornClient.cs:225]: Maybe we'll add some data about digging effectiveness to the item data files, and read this information into the item type definition.

**Comment by DevBobcorn** [comment]
I'm gonna add `action_type` field and `tier` field to entries in `items-<version>.json`, and introduce a new `item_tiers.json` for defining these tiers just like follows:
```json
{
    "wood": {
        "level": 0,
        "durability": 59,
        "speed": 2.0,
        "attack_damage": 0.0
    },
    "stone": {
        "level": 1,
        "durability": 131,
        "speed": 4.0,
        "attack_damage": 1.0
    },
    "iron": {
        "level": 2,
        "durability": 250,
        "speed": 6.0,
        "attack_damage": 2.0
    },
    "diamond": {
        "level": 3,
        "durability": 1561,
        "speed": 8.0,
        "attack_damage": 3.0
    },
    "gold": {
        "level": 0,
        "durability": 32,
        "speed": 12.0,
        "attack_damage": 0.0
    },
    "netherite": {
        "level": 4,
        "durability": 2031,
        "speed": 9.0,
        "attack_damage": 4.0
    }
}
```

**Comment by DevBobcorn** [comment]
Applicable blocks for mining/stripping/flattening/tilling will be defined in `block_interaction.json` as you suggested.

**Comment by Sieluna** [comment]
> I'm gonna add `action_type` field and `tier` field to entries in `items-<version>.json`, and introduce a new `item_tiers.json` for defining these tiers just like follows:
> 
> ```json
> {
>     "wood": {
>         "level": 0,
>         "durability": 59,
>         "speed": 2.0,
>         "attack_damage": 0.0
>     },
>     "stone": {
>         "level": 1,
>         "durability": 131,
>         "speed": 4.0,
>         "attack_damage": 1.0
>     },
>     "iron": {
>         "level": 2,
>         "durability": 250,
>         "speed": 6.0,
>         "attack_damage": 2.0
>     },
>     "diamond": {
>         "level": 3,
>         "durability": 1561,
>         "speed": 8.0,
>         "attack_damage": 3.0
>     },
>     "gold": {
>         "level": 0,
>         "durability": 32,
>         "speed": 12.0,
>         "attack_damage": 0.0
>     },
>     "netherite": {
>         "level": 4,
>         "durability": 2031,
>         "speed": 9.0,
>         "attack_damage": 4.0
>     }
> }
> ```

Actually, the `speed` is defined in `minecraft:tool`, but we are unable to parse it and `attack_damage` is defined at same place.
https://github.com/DevBobcorn/CornCraft/blob/1f36bbdfbb5c0f654dc10f1ba7c29b8ba2937314/Extra%20Data/items/items-1.20.6.json#L9055-L9068

**Comment by DevBobcorn** [comment]
1.20.5+ versions are using new data components to store this information, but it wasn't the case before that. Just checked vanilla source code and tiers are still used in 1.20.5+ to provide default values for these data components.

**Comment by DevBobcorn** [comment]
Maybe I'll remove `items-1.20.6.json` for now since I'm not sure how to properly export component-based item data, and it's still gonna be some time until we start working on support for 1.20.5+ versions.

**Comment by Sieluna** [comment]
@DevBobcorn

I have a different perspective on the current data structure. Due to the unique nature of tier, we can follow Mojang's approach and separate the tier level entirely. For example:
```java
"mineable": {
  "axe": {
    "minecraft:oak_planks": 0
  },
  "hoe": {
    "minecraft:oak_leaves": 0
  },
  "pickaxe": {
    "minecraft:stone": 1
  },
  "shovel": {
    "minecraft:dirt": 0
  }
}
```
This way, we don't need to modify the existing structure, as the tier level will essentially only be used here once.

**Comment by Sieluna** [comment]
The final result will looks like:
```csharp
private void PrepareMineableData(Json.JSONData interactions, BlockStatePalette palette)
{
    if (interactions.Properties.TryGetValue("mineable", out var mineableProperty))
    {
        var entries = mineableProperty.Properties;
        foreach (var (toolName, value) in entries)
        {
            var blocks = value.Properties;
            foreach (var (blockName, level) in blocks)
            {
                var blockId = ResourceLocation.FromString(blockName);
                if (palette.TryGetAllNumIds(blockId, out int[] stateIds))
                {
                    foreach (var stateId in stateIds)
                    {
                        if (mineableInteractionTable.TryGetValue(stateId, out var tools))
                        {
                            tools.Add(...ToolMap[toolName]);
                        }
                        else
                        {
                            mineableInteractionTable.Add(stateId, new List<int> { ...ToolMap[toolName] });   
                        }
                    }
                }
            }
        }
    }
}
 ```
 
 ToolMap is a tool relation like axe -> [netherite_axe:727, diamond_axe:722, ...]

**Comment by DevBobcorn** [comment]
Well the tier data is actually also used by weapon items (for calculating extra attack damage), and maybe keeping the tier names will make things more clear.

**Comment by DevBobcorn** [comment]
So I'm thinking maybe we make the structure this way:
```json
"mineable": {
  "axe": {
    "minecraft:oak_planks": "wood"
  },
  "hoe": {
    "minecraft:oak_leaves": "wood"
  },
  "pickaxe": {
    "minecraft:stone": "stone"
  },
  "shovel": {
    "minecraft:dirt": "wood"
  }
}
```

Might be a few more bytes in file size, but I think string ids/names are more readable and less likely to change than numeral ids.

**Comment by DevBobcorn** [comment]
Break Speed
So, when breaking a block with the right type of tool with tier higher than or equal to the minimum harvest tier(lime cells in the table), you get the tier speed multiplier along with an extra multiplier? How is that calculated?

**Comment by DevBobcorn** [comment]
Oh I found it, the multipiers was just two lines above the table. My bad :P

**Comment by DevBobcorn** [comment]
Just pushed updated data files and code, you can use the item action type and tier to determine mining time. (Entity effects are not implemented yet though)

**Comment by Sieluna** [comment]
There are still some missing features to be implemented, but it should work.

https://github.com/user-attachments/assets/d6e591d7-a6b4-4e4a-a632-14bb169c402a

**Comment by Sieluna** [comment]
```java
// case 1: break target drop target
this.dropSelf(Blocks.DIRT);

// case 2: break target drop other
this.dropOther(Blocks.FARMLAND, Blocks.DIRT);

// case 3: break target drop ? based on skill touch
this.add(Blocks.GRASS_BLOCK, block -> BlockLoot.createSingleItemTableWithSilkTouch(block, Blocks.DIRT));
this.add(Blocks.BOOKSHELF, block -> BlockLoot.createSingleItemTableWithSilkTouch(block, Items.BOOK, ConstantIntValue.exactly(3)));

//case 4: drop ?
this.dropPottedContents(Blocks.POTTED_OAK_SAPLING);

// case 5: slab
this.add(Blocks.ACACIA_SLAB, BlockLoot::createSlabItemTable);

// case 6: multipart target, break one drop more
this.add(Blocks.ACACIA_DOOR, block -> BlockLoot.createSinglePropConditionTable(block, DoorBlock.HALF, DoubleBlockHalf.LOWER));
```

// case 7: drop multi trigger by state:
this.add(Blocks.COCOA, block -> LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantIntValue.exactly(1)).add((LootPoolEntryContainer.Builder)BlockLoot.applyExplosionDecay(block, LootItem.lootTableItem(Items.COCOA_BEANS).apply((LootItemFunction.Builder)SetItemCountFunction.setCount(ConstantIntValue.exactly(3)).when(LootItemBlockStatePropertyCondition.hasBlockStateProperties(block).withProperty(CocoaBlock.AGE, 2)))))));

LootItemBlockStatePropertyCondition.Builder builder2 = LootItemBlockStatePropertyCondition.hasBlockStateProperties(Blocks.WHEAT).withProperty(CropBlock.AGE, 7);
this.add(Blocks.WHEAT, block -> BlockLoot.createCropDrops(block, Items.WHEAT, Items.WHEAT_SEEDS, builder2));

// case 8: skill touch ?= case 2?
this.dropWhenSilkTouch(Blocks.GLASS);

// case 9: no drop
this.add(Blocks.CAKE, BlockLoot.noDrop());

**Comment by DevBobcorn** [comment]
Does item drop logic need to be handled on clientside?

**Comment by Sieluna** [comment]
> Does item drop logic need to be handled on clientside?

Good question, I'm not sure. But I think we need to make sure the item is harvestable and the block is destroyed once we submit the mining completion command.

**Comment by Sieluna** [comment]
Do you have any plans for the item system? Personally, I'm not a fan of ScriptableObjects. We could actually stick with JSON using Unity’s serialization package. https://docs.unity3d.com/Packages/com.unity.serialization@3.1/manual/index.html

By the way, you can enable the "Discussion" feature, PR page is too messy.

**Comment by DevBobcorn** [comment]
> We could actually stick with JSON using Unity’s serialization package.

Sounds like a good idea. Those weapon item ScriptableObjects were just some ad-hoc dummies made quite some time ago for testing rigged player controllers, since we are implementing these for real those can be removed.



> By the way, you can enable the "Discussion" feature, PR page is too messy.

I'll check it out.

**Comment by DevBobcorn** [comment]
Discussion section is enabled now ˃ ᵕ ˂

**Comment by Sieluna** [comment]
I think this PR has too many changes, and reviewing it has become a bit tiring. How about we open a few other PRs to address the animation & effects part of the issue?

**Comment by DevBobcorn** [comment]
I'll merge it now so that we can start working on other things based on these changes.

**Pull request merged**